contracts/market/FifoQueue.sol

```diff
@@ -0,0 +1,57 @@
+
+pragma solidity ^0.4.24 ;
+import "../../node_modules/zeppelin-solidity/contracts//math/SafeMath.sol";
+
+contract FifoQueue {
+  using SafeMath for uint256; //todo jaycen PRELAUNCH - make sure we use this EVERYWHERE its needed
+
+  mapping(uint256 => uint256) fifoQueue;
+  uint256 first = 1;
+  uint256 last = 0;
+
+  function push(uint256 tokenId) public {
+    last = last.add(1);
+    fifoQueue[last] = tokenId;
+  }
+
+  function pop() public returns (uint256 tokenId) {
+    require(
+      last >= first,
+      "There are no items in the queue"
+    );
+
+    for(uint i = last; i < first; i = i.sub(1)) {
+      tokenId = peek();
+      if (tokenId == 0x0) {
```

Comment:
Here's where we check to see if the element in the mapping still exists (ie: the element hasn't been removed after it was pushed).